File: Classes/Extensions/UIKit/UIDevice+Info.m
//
//  UIDevice+Info.m
//  iSub
//
//  Created by Benjamin Baron on 12/6/20.
//  Copyright © 2020 Ben Baron. All rights reserved.
//

#import "UIDevice+Info.h"
#import "UIApplication+Helper.h"
#import <sys/sysctl.h>

@implementation UIDevice (Info)

+ (BOOL)isPad {
    return UIDevice.currentDevice.userInterfaceIdiom == UIUserInterfaceIdiomPad;
}

+ (BOOL)isSmall {
    // Should match iPhone SE 2nd Edition and iPhone 6/7/8 (all phones with home buttons)
    CGSize screenSize = UIScreen.mainScreen.bounds.size;
    return (UIInterfaceOrientationIsPortrait(UIApplication.orientation) ? screenSize.height : screenSize.width) < 700;
}

+ (NSString *)platform {
    // Get size
    size_t size;
    sysctlbyname("hw.machine", NULL, &size, NULL, 0);
    
    // Get value
    char *value = malloc(size);
    sysctlbyname("hw.machine", value, &size, NULL, 0);
    NSString *platform = [NSString stringWithCString:value encoding: NSUTF8StringEncoding];
    free(value);
    return platform;
}

+ (NSString *)systemBuild {
    int mib[2] = { CTL_KERN, KERN_OSVERSION };
    
    // Get size
    size_t size = 0;
    sysctl(mib, 2, NULL, &size, NULL, 0);
    
    // Get value
    char *value = malloc(size);
    sysctl(mib, 2, value, &size, NULL, 0);
    NSString *systemBuild = [NSString stringWithCString:value encoding: NSUTF8StringEncoding];
    free(value);
    return systemBuild;
}

+ (NSString *)completeVersionString {
    return [NSString stringWithFormat:@"%@ %@ (%@)", UIDevice.currentDevice.systemName, UIDevice.currentDevice.systemVersion, self.systemBuild];
}

@end

File: Classes/Extensions/UIKit/UIDevice+Info.swift
//
//  UIDevice+Info.swift
//  iSub
//
//  Created by Benjamin Baron on 1/27/21.
//  Copyright © 2021 Ben Baron. All rights reserved.
//

import Foundation

@objc extension UIDevice {
    static var isPad: Bool {
        Self.current.userInterfaceIdiom == .pad
    }
    
    static var isSmall: Bool {
        // Should match iPhone SE 2nd Edition and iPhone 6/7/8 (all phones with home buttons)
        let size = UIScreen.main.bounds.size
        let length = UIApplication.orientation.isPortrait ? size.height : size.width
        return length < 700
    }
    
    static func sysctl(name: String) -> String {
        var size = 0
        sysctlbyname(name, nil, &size, nil, 0)
        var value = [CChar](repeating: 0,  count: size)
        sysctlbyname(name, &value, &size, nil, 0)
        return String(cString: value)
    }
    
    // The device model identifier e.g. "iPhone12,5"
    static var deviceModel: String {
        sysctl(name: "hw.machine")
    }
    
    // The kernel version number e.g. "18C66"
    static var kernelVersion: String {
        sysctl(name: "kern.osversion")
    }
    
    // The complete OS version string e.g. "iOS 14.3 (18C66)"
    static var completeOSVersion: String {
        "\(Self.current.systemName) \(Self.current.systemVersion) (\(kernelVersion))"
    }
}
